### PR TITLE
Add toon shader node RPRBLND-1709

### PR DESCRIPTION
**Pull request by bsavery**
### PURPOSE
Adds a toon shader node for users

### EFFECT OF CHANGE
Users can create toon shader looks.  The node has a input color, or an advanced "ramp" style color interpolation

### CODE REVIEW

File: src/rprblender/nodes/rpr_nodes.py [hunk 1]
```diff
@@ -1513,3 +1513,95 @@ def export_hybrid(self):
...
+
+class RPRShaderNodeToon(RPRShaderNode):
+    ''' A toon shader using both the RPR Toon Shader and Ramp node '''
+    bl_label = 'RPR Toon'
+
+    def advanced_changed(self, context):
+        ramp_sockets = ['Shadow Color', "Mid Level","Mid Color", "Highlight Level", "Highlight Color"]
+        mix_sockets = ["Mid Level mix", "Highlight Level mix"]
+        if self.show_advanced:
+            self.inputs['Color'].enabled = Fals
```
Comment (ardraeiss) [inline src/rprblender/nodes/rpr_nodes.py:1526]: Typo, breaks node initialization, should be "False".

File: src/rprblender/nodes/rpr_nodes.py [hunk 2]
```diff
@@ -1513,3 +1513,95 @@ def export_hybrid(self):
...
+                interpolation_mode = pyrpr.INTERPOLATION_MODE_LINEAR if self.node.show_mix_levels \
+                    else pyrpr.INTERPOLATION_MODE_NONE
+                ramp = self.create_node(pyrpr.MATERIAL_NODE_TOON_RAMP, {
+                    pyrpr.MATERIAL_INPUT_SHADOW: self.get_input_value('Shadow Color'),
+                    pyrpr.MATERIAL_INPUT_MID: self.get_input_value('Mid Color'),
+                    pyrpr.MATERIAL_INPUT_HIGHLIGHT: self.get_input_value('Highlight Color'),
+                    pyrpr.MATERIAL_INPUT_POSITION1: self.get_input_value('Mid Level'),
+                    pyrpr.MATERIAL_INPUT_POSITION2: self.get_input_value('Highlight Level'),
+                    pyrpr.MATERIAL_INPUT_RANGE1: self.get_input_value('Mid Level Mix'),
+                    pyrpr.MATERIAL_INPUT_RANGE2: self.get_input_value('Highlight Level Mix'),
```
Comment (ardraeiss) [inline src/rprblender/nodes/rpr_nodes.py:1591]: Wrong input names.

File: src/rprblender/nodes/rpr_nodes.py [hunk 3]
```diff
@@ -1513,3 +1513,95 @@ def export_hybrid(self):
...
+            self.inputs['Color'].enabled = True
+
+    show_advanced: BoolProperty(name="Advanced", default=False, update=advanced_changed)
+    show_mix_levels: BoolProperty(name="Mix Levels", default=False, update=advanced_changed)
+
+    def init(self, context):
+        # Adding input sockets with default_value or hide_value properties.
+        # Here we use Blender's native node sockets
+        self.inputs.new('NodeSocketColor', "Color").default_value = (0.8, 0.8, 0.8, 1.0)    # Corresponds to Cycles diffuse
+        self.inputs.new('NodeSocketFloatFactor', "Roughness").default_value = 1.0
```
Comment (ardraeiss) [inline src/rprblender/nodes/rpr_nodes.py:1543]: Lack of soft/hard limits, editing this input in UI with mouse sets value to unreasonable numbers, for example "-3997".

**Review by ardraeiss — CHANGES_REQUESTED** [review]
UI and export doesn't work correctly

File: src/rprblender/nodes/rpr_nodes.py [hunk 4]
```diff
@@ -1513,3 +1513,95 @@ def export_hybrid(self):
...
+                    pyrpr.MATERIAL_INPUT_POSITION1: self.get_input_value('Mid Level'),
+                    pyrpr.MATERIAL_INPUT_POSITION2: self.get_input_value('Highlight Level'),
+                    pyrpr.MATERIAL_INPUT_RANGE1: self.get_input_value('Mid Level Mix'),
+                    pyrpr.MATERIAL_INPUT_RANGE2: self.get_input_value('Highlight Level Mix'),
+                    pyrpr.MATERIAL_INPUT_INTERPOLATION: interpolation_mode,
+                })
+
+                return self.create_node(pyrpr.MATERIAL_NODE_TOON_CLOSURE, {
+                    pyrpr.MATERIAL_INPUT_COLOR: (1.0, 1.0, 1.0, 1.0),
+                    pyrpr.MATERIAL_INPUT_NORMAL: self.get_input_link('Normal'),
```
Comment (ardraeiss) [inline src/rprblender/nodes/rpr_nodes.py:1597]: The None value(unlinked socket) causes error here. Something like that would be better:

            normal = self.get_input_link('Normal')
            if self.node.show_advanced:
            ...
                node = self.create_node(pyrpr.MATERIAL_NODE_TOON_CLOSURE, {
                    pyrpr.MATERIAL_INPUT_COLOR: (1.0, 1.0, 1.0, 1.0),
                    pyrpr.MATERIAL_INPUT_ROUGHNESS: self.get_input_value('Roughness'),
                    pyrpr.MATERIAL_INPUT_DIFFUSE_RAMP: ramp
                })

            else:
                node = self.create_node(pyrpr.MATERIAL_NODE_TOON_CLOSURE, {
                    pyrpr.MATERIAL_INPUT_COLOR: self.get_input_value('Color'),
                    pyrpr.MATERIAL_INPUT_ROUGHNESS: self.get_input_value('Roughness')
                })

            if normal:
                node.set_input(pyrpr.MATERIAL_INPUT_NORMAL, normal)

            return node

**Review by ardraeiss — CHANGES_REQUESTED** [review]

**Comment by bsavery** [comment]
Sorry @ardraeiss thanks for the review, I shouldn't have tagged as ready yet ;)

File: src/rprblender/nodes/rpr_nodes.py (hunk 3 above)
Comment (bsavery) [reply src/rprblender/nodes/rpr_nodes.py:1543]: BTW rpr diffuse uses this.

**Comment by ardraeiss** [comment]
Fixed forgotten names

**Review by ardraeiss — APPROVED** [review]

**Review by bnagirniak — APPROVED** [review]
Code is ok, but we are waiting for new core in #255

**Pull request merged**